### PR TITLE
Adding new useMyDomain flow to onboarding

### PR DESCRIPTION
**Pull request by delputnam**
#### Changes proposed in this Pull Request

This PR updates the onboading `use-your-domain` step to use the new flow already in use from the domain suggestions pages.

When selecting either the "Do you own it?" or the "Use a domain I own" links, the user will now be taken to the new steps to enter a domain and select whether they want to connect the domain or transfer it.

#### Testing instructions

Make sure that the flow works from both the `/start/domains` page and from the `/domains/add/use-my-domain` page (where it is already in use).

The pages should look something like this (depending on whether the selected domain is transferrable or not):
Screen Shot 2021-09-22 at 1 18 27 PM

Screen Shot 2021-09-22 at 1 18 39 PM

Here is a screen recording showing selecting the domain for transfer:

https://user-images.githubusercontent.com/1379730/134394550-4e6249a8-cb9f-4d2e-98ab-9c87a9a07386.mov

And here is a screen recording showing selecting the domain for connecting (mapping):

https://user-images.githubusercontent.com/1379730/134394649-d0bd7137-6b41-4a36-88be-fabb5773a79e.mov

### CODE REVIEW

**Review by rafaelgallani — CHANGES_REQUESTED** [review]
Nice job @delputnam! 
When testing, I found some things that differ from your previews/screenshots.

When selecting either the "Do you own it?" or the "Use a domain I own" options, I couldn't see the "Use a domain I own" header.

https://user-images.githubusercontent.com/18705930/134406957-41abcec1-c664-4276-a4e6-53b0028ec6e6.mov

I didn't check it too far, but I think that apparently it's somehow related to the `showHeader` prop on the `UseMyDomain` component... ? 

It's currently set as `false`: 

https://github.com/Automattic/wp-calypso/blob/b52ff3da3b3edf4295245e93bac9cc1adee9b31e/client/signup/steps/domains/index.jsx#L624

Setting it as `true` does this:

https://user-images.githubusercontent.com/18705930/134407389-2eac3eeb-293a-40b1-92f6-ad844685579b.mov

Apart from that, it looks good to me codewise 😄

**Comment by delputnam** [comment]
Thanks @rafaelgalani .  Looks like I missed pushing a commit. I'll have to go back and update this. :-/

**Comment by delputnam** [comment]
> Thanks @rafaelgalani . Looks like I missed pushing a commit. I'll have to go back and update this. :-/

FYI...missing commit has been pushed.

File: client/signup/steps/domains/style.scss
Comment (rafaelgallani) [inline client/signup/steps/domains/style.scss]:
```suggestion
	& .use-my-domain,
	& .connect-domain-step,
	& .domain-transfer-or-connect__content {
		background: transparent;
		box-shadow: none;
```

Comment (rafaelgallani) [reply client/signup/steps/domains/style.scss]: That will fix both the background color and the `box-shadow` prop on the transfer option which I referred to in the review.

**Review by rafaelgallani — APPROVED** [review]
Thanks, now everything is working as it should. 
Although I noticed some minor styling nitpicks that I haven't before. Sorry for that 😓 

- The component has a different background color than the onboarding background:

  image

  image

- Also, the `box-shadow` wasn't removed from the transfer option, which makes its style different than the previous step:

  image

  image

---

I added a fix to both of these in this review's comments. 😄

**Comment by delputnam** [comment]
Style update done.

**Review by rafaelgallani — APPROVED** [review]
LGTM, thanks @delputnam!

**Review by niranjan-uma-shankar — CHANGES_REQUESTED** [review]
Request a few changes:

1. The "Use a domain I own" page is not fully translated. This bug exists on the `/domains/add/use-my-domain/` route too. There's a specific sequence in which this bug happens. If you normally go through the signup flow, then the page appears translated. Now at the "Use a domain I own" step, do a browser reload of the page, and then you'll see the untranslated strings.
Screenshot 2021-10-06 at 3 00 44 PM

2. In the "Use a domain .." search box, if I type "google.com" or "domain.com", then I get a blank page with a console error.


uadiogoogle

3. We could trim spaces when checking for a valid domain. For example, in the following screenshot, searching for "matt.com " shows an error.


Screenshot 2021-10-06 at 4 40 44 PM

4. The "Use a domain" step in the blue signup background needs to be fixed. The blue signup is currently used for the paid plan signup flows - `/start/personal`, `/start/premium` etc.

Screenshot 2021-10-06 at 4 43 15 PM
Screenshot 2021-10-06 at 4 43 33 PM

5. In the `/new` flow, the "Use a domain I own" step jumps to a different signup flow, and hence shows the plans step from a different flow. To further explain, if I go to /new and select a new domain, then the plans step looks like this:



https://user-images.githubusercontent.com/1269602/136205434-9febcc23-2280-4e19-8144-b78a448788ad.mov

If I select "Already own a domain", then the plans from a different flow is displayed:


https://user-images.githubusercontent.com/1269602/136205642-9a785b78-1a02-4b9b-a02c-12fc0ea06555.mov

6. Back button from the plans step does not take back to the Transfer/Map page but instead goes one level up to the search input page:


uadio

7. Similar to the previous point, "back" button from the Transfer/Map page does not go back to the search input page but to the domains step page:
uadio2

 8. I think we should be more specific that transfers are included only in annual plans and not monthly plans. Can we change "Included in paid plans" to "Included in annual paid plans"?

Screenshot 2021-10-06 at 5 16 19 PM

9. My personal take is that I found the following line could be explained better:
**Included in paid plans**
**₹1,260.00/year**

It says included in paid plans but there's a price written directly below it, so it was not clear to me what's included in paid plans, and what is the cost component referring to. We could probably say "Domain transfer is free with paid plans or ₹1,260.00/year otherwise.", or maybe show a crossed-out price like in the domains step. 


10. Not an issue with this PR but noting it for completeness, the checkout page does not highlight mapped domains on the right column. For example, in the following screenshot, "ross.com" is a transferred domain that is mentioned in the "Included with your purchase" section:
Screenshot 2021-10-06 at 5 27 35 PM

But if "ross.com" is a domain mapping, then it's not highlighted:

Screenshot 2021-10-06 at 5 28 17 PM

11. In mobile screen, the "X" arrow alignment is slightly off:
Screenshot 2021-10-06 at 5 09 19 PM

**Comment by rafaelgallani** [comment]
Hey @niranjan-uma-shankar! 
Thanks for your review. 

We have already worked concurrently on a few points you have commented on. 
I have rebased this PR on `trunk` and started working on your other suggestions, but I would like to discuss a few points that I still have some questions about:

>In the /new flow, the "Use a domain I own" step jumps to a different signup flow, and hence shows the plans step from a different flow. To further explain, if I go to /new and select a new domain, then the plans step looks like this (...)

>If I select "Already own a domain", then the plans from a different flow is displayed

Regarding these, this PR shows the same plans that are currently displayed (we didn't change the current behavior):

https://user-images.githubusercontent.com/18705930/137408293-e61c677c-403b-4b90-b24f-522cb6a20b2f.mov

Should we stick to your suggestion and change them, or is it OK to keep them like this?

If we should change them, then what would be the best way to do it? I tried wrapping our component with the `useStepNavigation` hook, but that didn't work so well.

>Similar to the previous point, "back" button from the Transfer/Map page does not go back to the search input page but to the domains step page:

I was going to ask the same thing about the back button - is there an action we can dispatch or a hook/HOC we can use in this case? Or should we just change the URL? 

--- 

As I mentioned, some of the previous points you have commented on were already addressed. I'll be pushing them later today. 
Thanks again 😄

**Comment by niranjan-uma-shankar** [comment]
> Regarding these, this PR shows the same plans that are currently displayed (we didn't change the current behavior):
> Should we stick to your suggestion and change them, or is it OK to keep them like this?

I missed to mention in this in my comment, but you're right - this PR doesn't introduce this bug but I thought of bringing it up since it's related to the "Use a domain I own" step and I wonder if we could add a fix for it with this PR. 
@simison could you please let us know if this is a bug and should be fixed? For context, please check point 5 in [this comment](https://github.com/Automattic/wp-calypso/pull/56366#pullrequestreview-772710702) which describes a bug in the `/new` flow when Transfer/map domain option is selected.


> I was going to ask the same thing about the back button - is there an action we can dispatch or a hook/HOC we can use in this case? Or should we just change the URL?

I think if you pass the correct back URL value, the `StepWrapper` component should handle this for you, right? For example, in the domains step the backUrl is defined here:

https://github.com/Automattic/wp-calypso/blob/06771724d8d093fcc10bbe092968f57e1ab3c8f8/client/signup/steps/domains/index.jsx#L747-L781

**Comment by rafaelgallani** [comment]
Changing the base branch to #57157 - when it's approved it'll be changed to `trunk` again 😄

**Comment by rafaelgalani** [comment]
Once the other related PRs mentioned are approved, most of the points mentioned in the reviews will be addressed.

**Comment by rafaelgallani** [comment]
Excluding the last comment on the pricing (https://github.com/Automattic/wp-calypso/pull/56366#issuecomment-945688046), all of the reviews' comments have been addressed. 
A few of them are on separate PRs, as they get merged we'll change the base on this PR to `trunk`.

File: client/components/domains/use-my-domain/utilities/get-transfer-free-text.js
Comment (a8ci18n) [inline client/components/domains/use-my-domain/utilities/get-transfer-free-text.js]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **24** times:
`translate( 'Included in annual plans' )` *ES Score: 9*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/new-use-a-domain-i-own-flow-to-onboarding)

**Comment by delputnam** [comment]
Ok.  I think you're addressed all of the comments on this one now @rafaelgalani.  @niranjan-uma-shankar if you agree, would you mind approving this one?

**Review by niranjan-uma-shankar — CHANGES_REQUESTED** [review]
I tested and looks like a few of the issues still remain. There were also a couple of new ones, and I'm listing the below:

* The blue signup flow though is still having UI issues:

Here's the `/start/premium` flow:
Screenshot 2021-10-26 at 3 32 21 PM

* The error message text in red could do with better contrast, otherwise it's hard to read:

Screenshot 2021-10-26 at 3 38 47 PM

* The "Use a domain I own" header is misaligned with the rest of the page in the blue signup flow. This should probably be center aligned so that it sits nicely with the main header "Let's get your site a domain".

Screenshot 2021-10-26 at 3 38 57 PM


* The "Included in annual paid plans" doesn't appear translated. I suggest adding the `String freeze` label and waiting for translations before merge. Also the string is missing in the translated version for the mapped domain section

Screenshot 2021-10-26 at 3 50 27 PM

* In a non-en locale, the strings appear untranslated if you do the following steps:
   (1) go back from the Transfer/Map page and the "Use a domain I own" page appears untranslated 
   (2) click the Next button and Transfer/Map page is also untranslated

https://user-images.githubusercontent.com/1269602/138877615-f60188ae-e814-4aab-aa9e-ad58d64ef0ac.mov

* The page load after clicking the "Back" button from the Transfer/Map page felt a bit clunky. I haven't looked into the code yet, but the Back button in all the other signup steps feel responsive and quick, but in this case looks like a page reload is happening?

**Comment by rafaelgallani** [comment]
@niranjan-uma-shankar Thanks for the review!

>- The blue signup flow though is still having UI issues:

Sorry for that. My browser (Chrome) had the dark-mode flag enabled. So I wasn't viewing the same as you - and I thought that was already handled.

I added some changes to the styles. Here's how it's looking now:

- Text color and overall styles:
  
  image

- Error messages:
  
  image

As for the translation text, I opted to follow the bot's suggestion and use the `Included in annual plans` wording since this one is already translated and has already been used in some other domain-related places.

---

I couldn't reproduce the page-reloading case you got for non-en locales - what exact setup did you have on that scenario? 
I tried using both `pt-br` and `fr` languages - I had them set as primary on both `/me` and `/settings/general/:site`.

Although I think I have a clue: 

I can see that your path ends with `/fr`, and apparently, the reload happened because the locale was removed _after_ clicking the back button - which is probably happening because I didn't include the locale in the step URL. 
I added a commit that should address this.

Can you retake a look at it? If it's still not solved, then could you point me on how to get to this state?

**Comment by rafaelgallani** [comment]
Rebasing on `trunk` since #57157 has been merged. _(the second rebase was just a `squash`ed version, simplifying some commits 😄)_

**Review by niranjan-uma-shankar — COMMENTED** [review]
Thanks for making these changes! I confirm that most have been fixed. 

> Can you retake a look at it? If it's still not solved, then could you point me on how to get to this state?

The page reload issue is now fixed 👍 . When I go back from the Transfer/Map page, the page no longer reloads and the translations are in place.

> As for the translation text, I opted to follow the bot's suggestion and use the Included in annual plans wording since this one is already translated and has already been used in some other domain-related places.

Sounds good to me. 

> Error messages:

I still found this contrast to be hard to read, and this has accessibility implications. I checked using the [contrast checker tool](https://webaim.org/resources/contrastchecker/) and [here's the screenshot](https://cloudup.com/cs12OrhviIm) of the report. _I don't consider this to be a blocker_, so feel free to follow up in a separate PR. 

> I added some changes to the styles. Here's how it's looking now:

Thanks, this is looking good in the blue signup flow.

A few more things I observed. I think if (1) and (2) below can be fixed, then we should be good to merge. For the other issues I've noted below, it would be nice to track it as GH issues if you'd like to come to it later. 

1. The "Included in annual plans" string does not appear for the "Connect your domain" part if it's a new signup. However, it shows on page reload. Check the recording below which shows the missing string, which then shows up on page reload:


https://user-images.githubusercontent.com/1269602/139009008-2f45f755-afa7-4678-bd5a-240ef89d5e91.mov

2. In the launch site flow, the Tranfer/Map page does not get translated. I see this issue only in the launch flow, and not in the regular signup flow:

https://user-images.githubusercontent.com/1269602/139009443-31c77a12-64b1-4e7e-86f1-562b2f07128d.mov



3. _Not a blocker_. In the signup flow, page reload on the Transfer/Map page does not preserve the translations. 

https://user-images.githubusercontent.com/1269602/139009326-3d53191a-b176-46da-8a26-6b76b21f9064.mov

4. _Not a blocker_. If I use a ccTLD, then I'm initially shown the option to only connect and not transfer, but if I go back to the same screen, the option disappears.

https://user-images.githubusercontent.com/1269602/139010540-b0e7a755-3203-4e94-a28a-c1ac13e4f0d5.mov

5. _Not a blocker_. If you enter the `/new` flow logged out, then the "Already own a domain" option does not appear. This is not a priority right now since `/new` flow is only visible for existing logged in users but in the future it could be made visible to logged out users too, so I'm noting it here in case you'd like to fix it in a follow up. 
Screenshot 2021-10-27 at 10 43 58 AM

6. _Not a blocker_. If I enter the `/new` flow as a logged in user in `en` locale by clicking on the "Add new site" option, then use the language picker to switch language, the "Use a domain" step does not pick up the translations. In the recording below, I switched to `es` and yet got english strings in the transfer step.

https://user-images.githubusercontent.com/1269602/139011452-cdae5231-1df1-4020-a27c-a4cf384a3369.mov

7. _Not a blocker_. I'd noted this in an earlier comment, but we could maybe trim spaces in the input box, so that a string like "matt.com " [does not show an error](https://cloudup.com/c6iGgXjvTNZ).

**Comment by rafaelgallani** [comment]
@niranjan-uma-shankar I believe that most of the issues you are seeing - _both blockers and non-blockers_ - will be fixed as soon as #57064 is merged. I'll rebase after that again and you can take one final look at it.

**Comment by niranjan-uma-shankar** [comment]
Sure, please ping me when it's ready and I'll take a look again. There's a unit test failing on this PR which would need to get fixed.

**Review by niranjan-uma-shankar — COMMENTED** [review]
Another non-blocking bug which I forgot to mention in my previous comment. The Back button on the domain step does not take back to Customer Home page. To reproduce:
* Enter the site launch flow. Notice that the domain step has a "Back to my sites" button and clicking this will take you to Customer Home. 
* Click on "Use a domain I own"
* Type a domain in the search box 
* In the Transfer/Map page, click Back, and then click Back one more time for the "Use a domain I own" page to get to the domains step.
* Notice that the domains step no longer has a "Back to my sites" button but instead has a Back button. Clicking it does not take to Customer Home but instead to the "Use a domain .. " step. 

Launch flow with "Back to my sites" button

https://user-images.githubusercontent.com/1269602/139256471-9f855f3d-50e2-485b-b2b3-7269a137096a.mov

Launch flow with Back button that does not take to Customer Home:



https://user-images.githubusercontent.com/1269602/139256514-66661c6b-fec2-4547-b9f0-411f5d9f0ed8.mov

**Comment by rafaelgallani** [comment]
Rebasing with latest `trunk` changes

**Comment by rafaelgallani** [comment]
@niranjan-uma-shankar If you could please have one last look at it, all the related PRs have been merged now. 
All of the problems you reported - except the 6th and 7th, which we will handle in a follow-up PR since it's currently like this and it's not a blocker - should be fixed now.

**Review by niranjan-uma-shankar — APPROVED** [review]
Thanks for working through the changes @rafaelgalani. I'm approving this PR but there are still a few issues. I'm grouping them into "good to fix in this PR" and "can be followed up". For the former, feel free to take the final call, and if you agree with my recommendation you can push in the fixes and merge without waiting for my review again(I'm also happy to review it if you'd like). For the latter, you could maybe track it somewhere on your project board so that someone can get to them later.

**Good to fix in this PR**

1. Point number 1 in [this comment](https://github.com/Automattic/wp-calypso/pull/56366#pullrequestreview-790207171) is not fixed:
> The "Included in annual plans" string does not appear for the "Connect your domain" part if it's a new signup. However, it shows on page reload. Check the recording below which shows the missing string, which then shows up on page reload:

**Can be followed up**

1. In https://github.com/Automattic/wp-calypso/pull/57288, the Back button was removed in  mobile web, and the only way to go back to previous steps is to use the browser's back button. So, when I clicked the browser back button on the Transfer/map page, there was a page reload and a weird error state that the domain cannot be transferred or connected. Check the recording below. 

https://user-images.githubusercontent.com/1269602/139639513-3adbb676-4168-4a38-b474-ed2611e0a971.mov

2. In the following video, you can see that going back from plans step fires a search with a different domain and an error state is displayed that the domain cannot be transferred or connected. What I did - (a) Search for "Jeff's bagels" in the domain step (2) Click on Use a domain .. link and entered a domain jeffbagel.com (3) Selected transfer option (4) Hit the Back button on Plans step. At this point, instead of showing jeffbagel.com, Tranfer/Map option showed the result for "jeff's bagels", which was the search term in the domain step and reports an error. I would've expected it to use the search term I used in the Use a domain step instead of the domain step.


https://user-images.githubusercontent.com/1269602/139640416-bf288c27-ad44-45df-a739-c49482c90da8.mov

3. Point (5), (6), and (7) in [this comment](https://github.com/Automattic/wp-calypso/pull/56366#pullrequestreview-790207171).

4. Error contrast color that I observed in an earlier comment:
> I still found this contrast to be hard to read, and this has accessibility implications. I checked using the contrast checker tool and here's the screenshot of the report. I don't consider this to be a blocker, so feel free to follow up in a separate PR.

5. Point (5) from [this comment](https://github.com/Automattic/wp-calypso/pull/56366#pullrequestreview-772710702). You could probably discuss this with someone in Forge team to check if it's worth taking up. 
> In the /new flow, the "Use a domain I own" step jumps to a different signup flow, and hence shows the plans step from a different flow.

**Comment by niranjan-uma-shankar** [comment]
@rafaelgalani One more point - we have a couple of signup flow tests running, so could you please check with the folks in A8C Experiments if it's fine to deploy this change? Just to confirm that deploying this change will not affect the currently running experiments.

**Comment by rafaelgallani** [comment]
@niranjan-uma-shankar We will create internal tasks to handle the non-blocking "Can be followed up" issues separately. I double-checked and, currently, there are no running experiments on the signup flow. 

I'll be deploying this one. Thanks for all the thorough reviews 😄

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6945664

Thank you @delputnam for including a screenshot in the description! This is really helpful for our translators.

File: client/signup/steps/domains/index.jsx
Comment (tyxla) [inline client/signup/steps/domains/index.jsx:667]: Perhaps you meant to access `this.props.step` or `this.state.currentStep` here? I don't see a `step` component state being set anywhere.

File: client/signup/steps/domains/index.jsx
Comment (tyxla) [inline client/signup/steps/domains/index.jsx:681]: It is not a good time to perform any side effects or set state during rendering.

**Review by tyxla — COMMENTED** [review]
I see this has been merged already, but I just wanted to bring up two potential problems I noticed as I was glancing through the code.

File: client/signup/steps/domains/index.jsx
Comment (rafaelgallani) [reply client/signup/steps/domains/index.jsx:681]: @tyxla Thanks for catching that! This code was for testing purposes (only), and I forgot to remove it. 
It's not doing anything currently. I'll create a PR for it 😄 

Edit: addressed in #57753.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6945664).

**Pull request merged**
